### PR TITLE
SAK-43649: SWITCH: Permissions widget: Stay in page after Undo Changes

### PR DESCRIPTION
**Pull request by frasese**

### CODE REVIEW

**Comment by frasese** [comment]
@adrianfish please, can you take a look at this? I've had a lots of problems with the .checked property and finally I've had to modify the shadow DOM directly.

**Comment by steveswinsburg** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/sakaiproject/sakai/pullRequest?prid=5521027&bid=18088217)

File: webcomponents/tool/src/main/frontend/js/sakai-permissions.js [hunk 1]
```diff
@@ -65,6 +65,16 @@ class SakaiPermissions extends SakaiElement {
...
   updated(changedProperties) {
+    let inputs = this.renderRoot.querySelectorAll("input[type='checkbox']");
+    inputs.forEach(elem => { 
+      let role = elem.getAttribute('data-role');
+      let perm = elem.getAttribute('data-perm');
+      
+      if(role && perm){
+        elem.checked = this.on[role].includes(perm);
+      }
+    })
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/sakai-permissions.js:76]: I'd move this code into your resetPermissions function body. You do this by waiting on the updateComplete promise.

this.updateComplete.then(() => {
let inputs = this.renderRoot.querySelectorAll("input[type='checkbox']");
    inputs.forEach(elem => { 
      let role = elem.getAttribute('data-role');
      let perm = elem.getAttribute('data-perm');

      if (role && perm) {
        elem.checked = this.on[role].includes(perm);
      }
    })
});

... you get the idea. That way the  render has finished and you can manipulate the checks.

File: webcomponents/tool/src/main/frontend/js/sakai-permissions.js [hunk 2]
```diff
@@ -65,6 +65,16 @@ class SakaiPermissions extends SakaiElement {
   get tool() { return this._tool; }
 
   updated(changedProperties) {
+    let inputs = this.renderRoot.querySelectorAll("input[type='checkbox']");
+    inputs.forEach(elem => { 
+      let role = elem.getAttribute('data-role');
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/sakai-permissions.js:70]: blank line here please, it makes js easier to read, imo, if you leave a blank line at the start of any multi line function body, even lambdas.

File: webcomponents/tool/src/main/frontend/js/sakai-permissions.js [hunk 3]
```diff
@@ -65,6 +65,16 @@ class SakaiPermissions extends SakaiElement {
...
+      let perm = elem.getAttribute('data-perm');
+      
+      if(role && perm){
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/sakai-permissions.js:73]: Please leave a space between between if and the brace. Same for the curly brace on the end.
if (role && perm) {

I am very fussy about these tiny things ... :) I also like a blank line after the function definition if the body has more than one line of code in it. I am fussy, ask Miguel!

**Comment by adrianfish** [comment]
https://lit-element.polymer-project.org/guide/lifecycle#updatecomplete

File: webcomponents/tool/src/main/frontend/js/sakai-permissions.js (hunk 1 above)
Comment (frasese) [reply webcomponents/tool/src/main/frontend/js/sakai-permissions.js]: Nice! I like the idea. So, no need to reload or re-render. Sorry but, I'm still learning about webcomponents and litelement.

I think I've set all the blank lines and spaces as you like. Anyway, if anything else is needed, just say it.

Thanks for your help.

Comment (frasese) [reply webcomponents/tool/src/main/frontend/js/sakai-permissions.js]: The dispatchEvent part is to reset the table cell style arround the checkbox. 
We also can do this calling this.attachHandlers() or just:
`$('.permissions-table input:checkbox').change();`

But I don't like the idea to mix vanilla javascript with jquery (at least inside the same fragment/method)

File: webcomponents/tool/src/main/frontend/js/sakai-permissions.js [hunk 4]
```diff
@@ -172,6 +172,31 @@ class SakaiPermissions extends SakaiElement {
...
+
+          let role = elem.getAttribute('data-role');
+          let perm = elem.getAttribute('data-perm');
+          if (role && perm) {
+
+            elem.checked = this.on[role].includes(perm);
+            if ("createEvent" in document) {
+
+                let evt = document.createEvent("HTMLEvents");
+                evt.initEvent("change", false, true);
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/sakai-permissions.js:189]: Can't you just use "new Event("change", { bubbles: true })" for these lines? Maybe you don't even need to bubble it.

https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Creating_and_triggering_events

Comment (frasese) [reply webcomponents/tool/src/main/frontend/js/sakai-permissions.js]: Yes. You are right. I'm still attached to the old ways...

**Comment by steveswinsburg** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- webcomponents/tool/src/main/frontend/js/sakai-permissions.js  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/sakaiproject/sakai/pullRequest?prid=5521027&bid=18088217)

File: webcomponents/tool/src/main/frontend/js/sakai-permissions.js [hunk 5]
```diff
@@ -172,6 +172,30 @@ class SakaiPermissions extends SakaiElement {
...
+      inputs.forEach(elem => {
+
+          let role = elem.getAttribute('data-role');
+          let perm = elem.getAttribute('data-perm');
+          if (role && perm) {
+            let elemChanged = (elem.checked != this.on[role].includes(perm));
+            elem.checked = this.on[role].includes(perm);
+            if (elemChanged) {
+              if (elem.dispatchEvent) {
+                  elem.dispatchEvent(new Event("change"));
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/sakai-permissions.js:188]: I'd get rid of this test. Only really old crappy browsers don't have this. Up to you though.

**Comment by steveswinsburg** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Complexity increasing per file
==============================
- webcomponents/tool/src/main/frontend/js/sakai-permissions.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/sakaiproject/sakai/pullRequest?prid=5521027&bid=18088217)

**Comment by mpellicer** [comment]
Thanks for reviewing @frasese

**Pull request merged**